codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -0,0 +1,487 @@
+            ) {
+                rustTemplate(
+                    """
+                    let mut tokens_owned = #{json_token_iter}(input).peekable();
+                    let tokens = &mut tokens_owned;
+                    """,
+                    *codegenScope
+                )
+                rust("let result =")
+                deserializeMember(member)
+                rustTemplate(".ok_or_else(|| #{Error}::custom(\"expected payload member value\"));", *codegenScope)
+                expectEndOfTokenStream()
+                rust("result")
+            }
+        }
+    }
+
+    override fun operationParser(operationShape: OperationShape): RuntimeType? {
+        // Don't generate an operation JSON deserializer if there is no JSON body
+        val httpDocumentMembers = httpBindingResolver.responseMembers(operationShape, HttpLocation.DOCUMENT)
+        if (httpDocumentMembers.isEmpty()) {
+            return null
+        }
+
+        val outputShape = operationShape.outputShape(model)
+        val fnName = symbolProvider.deserializeFunctionName(operationShape)
+        return RuntimeType.forInlineFun(fnName, "json_deser") {
+            it.rustBlockTemplate(
+                "pub fn $fnName(input: &[u8], mut builder: #{Builder}) -> Result<#{Builder}, #{Error}>",
+                "Builder" to outputShape.builderSymbol(symbolProvider),
+                *codegenScope
+            ) {
+                rustBlock("if !input.is_empty()") {
+                    rustTemplate(
+                        """
+                        let mut tokens_owned = #{json_token_iter}(input).peekable();
+                        let tokens = &mut tokens_owned;
+                        #{expect_start_object}(tokens.next())?;
+                        """,
+                        *codegenScope
+                    )
+                    deserializeStructInner(httpDocumentMembers)
+                    expectEndOfTokenStream()
+                }
+                rust("Ok(builder)")
+            }
+        }
+    }
+
+    override fun errorParser(errorShape: StructureShape): RuntimeType? {
+        if (errorShape.members().isEmpty()) {
+            return null
+        }
+        val fnName = symbolProvider.deserializeFunctionName(errorShape) + "json_err"
+        return RuntimeType.forInlineFun(fnName, "json_deser") {
+            it.rustBlockTemplate(
+                "pub fn $fnName(input: &[u8], mut builder: #{Builder}) -> Result<#{Builder}, #{Error}>",
+                "Builder" to errorShape.builderSymbol(symbolProvider),
+                *codegenScope
+            ) {
+                rustTemplate(
+                    """
+                    let mut tokens_owned = #{json_token_iter}(input).peekable();
+                    let tokens = &mut tokens_owned;
+                    #{expect_start_object}(tokens.next())?;
+                    """,
+                    *codegenScope
+                )
+                deserializeStructInner(errorShape.members())
```

Comment:
does this read until `}`? we should make sure that extra key/value pairs are allowed/ignored

Reply:
Yes, `deserializeStructInner` consumes the final `}` as well as skips values for keys it doesn't recognize.